### PR TITLE
Handle JavaScript dialogs

### PR DESCRIPTION
**Pull request by padde**
Implements JS dialog handling, as discussed in https://github.com/keathley/wallaby/issues/169.

### CODE REVIEW

File: test/support/pages/alert.html [hunk 1]
```diff
@@ -1,21 +0,0 @@
...
-    function doPrompt() {
-      var name = prompt("name", "");
-        if (name != null) {
-            document.getElementById("result").innerHTML =
-            "Hello " + name;
-        }
-    }
-</script>
-  </body>
-</html>
```
Comment (padde) [inline test/support/pages/alert.html]: I felt free to rename and modify this fixture, since it wasn't actually used anywhere. Let me know if you're somehow attached to it or have other reasons to keep this as is 😉

**Comment by padde** [comment]
Hey @keathley I've got an initial version done, would you mind taking a look?

**Comment by keathley** [comment]
Thanks for this @padde! This looks pretty good to me. My only thought was that it might make sense to make `accept_alert` etc. chainable. I'm not sure if thats a good idea or not. Being able to see the message is definitely useful. I wonder if we could do that with another callback as the 3rd argument or if that api would just get messy. What do you think @PragTob?

File: lib/wallaby/browser.ex [hunk 2]
```diff
@@ -980,6 +980,63 @@ defmodule Wallaby.Browser do
     current_url(session) == "about:blank"
   end
 
+  @doc """
+  Accepts all subsequent JavaScript dialogs in the given session.
+  """
```
Comment (PragTob) [inline lib/wallaby/browser.ex]: Thanks for adding docs right away! :+1:

File: lib/wallaby/browser.ex [hunk 3]
```diff
@@ -980,6 +980,63 @@ defmodule Wallaby.Browser do
...
+  Dismisses all subsequent JavaScript dialogs in the given session.
+  """
+  def dismiss_dialogs(%Session{}=session) do
+    Driver.dismiss_dialogs(session)
+    session
+  end
+
+  @doc """
+  Accepts one alert dialog, which is triggered within the specified `fun`.
+  Returns the last message that was presented to the user.
```
Comment (PragTob) [inline lib/wallaby/browser.ex]: at least for the first one of those it might be nice to show a code example of what might be meant by `triggered by specified fun` :)

File: lib/wallaby/browser.ex [hunk 4]
```diff
@@ -980,6 +980,63 @@ defmodule Wallaby.Browser do
...
+  Dismisses one confirm dialog, which is triggered within the specified `fun`.
+  Returns the last message that was presented to the user.
+  """
+  def dismiss_confirm(%Session{}=session, fun) do
+    Driver.dismiss_confirm(session, fun)
+  end
+
+  @doc """
+  Accepts one confirm dialog, which is triggered within the specified `fun`.
+  Returns the last message that was presented to the user.
```
Comment (PragTob) [inline lib/wallaby/browser.ex]: Should mention that this is a prompt an the input value needs to be specified :)

File: lib/wallaby/browser.ex [hunk 5]
```diff
@@ -980,6 +980,63 @@ defmodule Wallaby.Browser do
...
+  """
+  def accept_prompt(%Session{}=session, options \\ [], fun) do
+    input_value = Keyword.get(options, :with)
+    Driver.accept_prompt(session, input_value, fun)
+  end
+
+  @doc """
+  Dismisses one confirm dialog, which is triggered within the specified `fun`.
```
Comment (PragTob) [inline lib/wallaby/browser.ex]: it's a prompt :)

File: test/support/pages/alert.html (hunk 1 above)
Comment (PragTob) [inline test/support/pages/alert.html]: :+1:

File: test/wallaby/browser/dialog_test.exs [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    {:ok, %{page: page}}
+  end
+
+  describe "accept_dialogs/1" do
+    test "accept window.alert", %{page: page} do
+      result = page
+      |> accept_dialogs()
+      |> click(Query.link("Alert"))
+      |> find(Query.css("#result"))
+      |> Element.text()
```
Comment (PragTob) [inline test/wallaby/browser/dialog_test.exs]: hm, this is mostly a style question and dunno how @keathley sees this but I prefer:

```
result = page
         |> thing
         |> other_thing
```

or 

```
result =
  page
  |> thing
  |> other_thing
```

Looks more tidy imo :)

**Comment by PragTob** [comment]
Oh @keathley - having it chainable would be great, although with accept_dialogs being chainable I'm not sure how important it is as you can just use that if you don't care and then use the more detailed methods if you care about the values and then it's not so bad that it isn't chainable.

I'd like them to be chainable, but two functions as arguments seems weird as interface for a function. :|

**Comment by padde** [comment]
Thanks for the comments guys, all valid points. Will push some improvements soon.

Regarding the chainable API issue let me summarize:

* `accept_alert(session, fun) -> string` (current solution)
  * not chainable
  * OTOH there is `accept_dialogs` for when we don't care about messages
* `accept_alert(session) -> session`, then `last_message(session) -> string`
  * easily implemented with PhantomJS (in fact the current solution internally does this)
  * not an option because it would be incompatible with the WebDriver spec, which only allows to get the message of the currently visible prompt *before* interacting with it
* `current_message(session) -> string`, then `accept_alert(session) -> session`
  * easily implemented with WebDriver
  * not an option because this is not possible with PhantomJS - it has no concept of "the current dialog". At the time when we know the message the alert is already accepted/dismissed
* `accept_alert(session, trigger_fun, message_callback_fun) -> session`
  * I agree having to supply two callback functions seems weird
  * I don't like having to trust the underlying code to actually invoke the assertion

On a more general note, I dislike the fact that the current solution mixes concerns by both interacting with the interface and querying it for the alert message. Alas, it seems near impossible to overcome the limitations of PhantomJS' implementation in this regard.

File: test/wallaby/browser/dialog_test.exs (hunk 6 above)
Comment (padde) [inline test/wallaby/browser/dialog_test.exs]: I actually prefer the style in the second example, but tried to stick to the way existing tests were written, because those that I had looked at were like this. As I read some more code and saw different styles being mixed, I'm not so sure which is the preferred style. Let @keathley decide 😉

File: lib/wallaby/browser.ex (hunk 5 above)
Comment (padde) [inline lib/wallaby/browser.ex]: Duh, good catch :)

File: lib/wallaby/phantom/driver.ex [hunk 7]
```diff
@@ -333,6 +333,135 @@ defmodule Wallaby.Phantom.Driver do
...
+    var page = this;
+    page.__onAlertDefault = page.onAlert;
+    page.onAlert = function(msg) {
+      page.__alertMessage = msg;
+      page.onAlert = page.__onAlertDefault;
+    }
+    return "ok";
+    """
+    {:ok, "ok"} = execute_phantom_script(session, script)
+    fun.(session)
```
Comment (padde) [inline lib/wallaby/phantom/driver.ex]: @keathley @PragTob what do you think about the session parameter being passed as argument to the callback function - do we really need this? Are there use cases where the session struct is changed through "normal usage" such that we need to make sure that we pass it along all the way?

File: test/wallaby/browser/dialog_test.exs (hunk 6 above)
Comment (padde) [inline test/wallaby/browser/dialog_test.exs]: I went with the the second version now, let me know if you had something else in mind @keathley.

**Comment by padde** [comment]
@keathley @PragTob everything should be fixed now. Your opinion on [this comment](https://github.com/keathley/wallaby/pull/173#discussion_r108791947) might result in additional changes though.

**Comment by padde** [comment]
Oh and maybe I should add a section to the README about handling JavaScript dialogs?

File: lib/wallaby/phantom/driver.ex (hunk 7 above)
Comment (PragTob) [inline lib/wallaby/phantom/driver.ex]: At first I thought we didn't need it, as we have it in context most of the time, but then I thought of a case were you wanna implement as a separate method like

```
defp my_fun(session), do: #....
````

and then pass it as `&my_fun/1` - as it needs access to the session `&my_fun/0` wouldn't work so I like this :)

Comment (padde) [inline lib/wallaby/phantom/driver.ex]: Fair point. We could make the parameter optional, something along the lines of

```elixir
callback =
  case :erlang.fun_info(fun) |> Keyword.get(:arity) do
    0 -> fn (_) -> fun.() end
    1 -> fun
    _ -> raise BadArityError, args: [session], function: fun
  end
Driver.accept_alert(session, callback)
```

Not sure if this would be worth it though.

Comment (PragTob) [inline lib/wallaby/phantom/driver.ex]: yeah me neither, I think it's fairly common for people to just do `fn(_) -> i_dont_need_your_arguments end`

**Comment by PragTob** [comment]
I think we could get this in now, thanks for the README section @padde - thoughts @keathley ? :)

**Comment by padde** [comment]
Nice! Any ideas when we can expect a release?

By the way: awesome experience working with you guys, shows once again what a wonderful community this is. Greatly appreciated, thanks!

**Comment by keathley** [comment]
@padde Thanks! I've added it to our 0.17.0 release milestone. Hopefully it won't take too long to get that out.

File: lib/wallaby/browser.ex [hunk 8]
```diff
@@ -1040,6 +1040,105 @@ defmodule Wallaby.Browser do
...
+  ```
+  message = accept_prompt session, [with: "User input"], fn(s) ->
+    click(s, Query.link("Trigger prompt"))
+  end
+  ```
+  """
+  def accept_prompt(%Session{}=session, options \\ [], fun) do
+    input_value = Keyword.get(options, :with)
+    Driver.accept_prompt(session, input_value, fun)
+  end
```
Comment (padde) [inline lib/wallaby/browser.ex]: - [ ] Maybe we could add a safeguard here in case the supplied `input_value` is not a string, because  `window.prompt` will always return a string except when it's dismissed. For example, providing an integer as argument could be considered an error. Likewise, passing `nil` would be equivalent to calling `dismiss_prompt`, so that function should be used instead.

@PragTob @keathley what do you think? Should we raise an error here? Or maybe a warning would be more appropriate?

File: lib/wallaby/phantom/driver.ex [hunk 9]
```diff
@@ -333,6 +333,135 @@ defmodule Wallaby.Phantom.Driver do
...
+    page.__onConfirmDefault = page.onConfirm;
+    page.onConfirm = function(msg) {
+      page.__confirmMessage = msg;
+      page.onConfirm = page.__onConfirmDefault;
+      return returnVal;
+    }
+    return "ok";
+    """
+    {:ok, "ok"} = execute_phantom_script(session, script, [return_value])
+    fun.(session)
```
Comment (padde) [inline lib/wallaby/phantom/driver.ex]: - [ ] When invoking the callbacks like here, do we need to catch errors to make sure that the callback is properly reset? IMO this would only be a problem if:

* inside the callback an error is raised *before* the dialog is triggered
* the test that actually uses this `accept_*`/`dismiss_*` function catches errors itself
* so when the error was catched in the surrounding test, the callback handler is in place, but was not reset back to the default handler

Might be an edge case that's not worth handling...

File: lib/wallaby/phantom/driver.ex [hunk 10]
```diff
@@ -333,6 +333,135 @@ defmodule Wallaby.Phantom.Driver do
...
+    script = """
+    var page = this;
+    page.onConfirm = page.__onConfirmDefault;
+    return ["ok", page.__confirmMessage];
+    """
+    {:ok, ["ok", message]} = execute_phantom_script(session, script)
+    message
```
Comment (padde) [inline lib/wallaby/phantom/driver.ex]: - [ ] Instead of returning the message here we could still keep retrieval of messages as a separate function and then make the `accept_*`/`dismiss_*` functions chainable .

This would of course require us to store the messages somewhere, most likely in the `Session` struct. That would of course require the user to actually store the new state back in the `session`/`page` variable within the test, not sure if it's worth that. AFAICS this is not necessary with all of the other browser  functions so far.

For example the following would work:

```
session =
  session
  |> visit("dialogs.html")
  |> accept_alert(fn (s) ->
    click(s, Query.link("Trigger alert")
  end)

message = last_dialog_message(session)
```

but this would not work:

```
session
|> visit("dialogs.html")
|> accept_alert(fn (s) ->
  click(s, Query.link("Trigger alert")
end)

message = last_dialog_message(session)
```

File: lib/wallaby/phantom/driver.ex [hunk 11]
```diff
@@ -333,6 +333,135 @@ defmodule Wallaby.Phantom.Driver do
...
+    return "ok";
+    """
+    {:ok, "ok"} = execute_phantom_script(session, script, [return_value, use_default])
+    fun.(session)
+    script = """
+    var page = this;
+    page.onPrompt = page.__onPromptDefault;
+    return ["ok", page.__promptMessage];
+    """
+    {:ok, ["ok", message]} = execute_phantom_script(session, script)
```
Comment (padde) [inline lib/wallaby/phantom/driver.ex]: By the way, this nested "fake :ok tuple" is necessary, because in case of a JavaScript error on the PhantomJS side, it will unfortunately give us a 200 OK response with the return value being a JS stack trace instead of the expected value. Maybe there's a better way to handle this somehow?

File: lib/wallaby/phantom/driver.ex (hunk 10 above)
Comment (padde) [inline lib/wallaby/phantom/driver.ex]: Poltergeist for example tracks the occurred dialog messages within the JS client, might be worth looking into: https://github.com/teampoltergeist/poltergeist/blob/f7a8e970562fd1f912b39a0acc1afe7761d16a30/lib/capybara/poltergeist/client/browser.coffee#L73-L75

Comment (padde) [inline lib/wallaby/phantom/driver.ex]: Oh now I realize that I didn't like that approach because it would not work the same way with other drivers. We'd have to keep track of the messages for other browsers within Wallaby if we'd choose this API. Sorry for the confusion :/

File: lib/wallaby/browser.ex (hunk 8 above)
Comment (PragTob) [inline lib/wallaby/browser.ex]: If it should always be a string, the easiest version (to my mind) would be to just have a `when` clause or something in `Driver.accept_prompt` that matches if it is a binary/string and otherwise either a match error or another implementation that raises ind is like "these are the arguments something went wrong"

File: lib/wallaby/phantom/driver.ex (hunk 9 above)
Comment (PragTob) [inline lib/wallaby/phantom/driver.ex]: It's good that you think of that, and it might be nice if we ensured that it is reset correctly and then reraise the error or something. I think that'd be the most correct thing to do :)

However, we can do that in a separate PR or even only if the problem arises. E.g. it should only affect the current session (or so I hope/guess) and when an error is raised from that function it is already broken either way so the impact shouldn't be high.

File: lib/wallaby/phantom/driver.ex (hunk 10 above)
Comment (PragTob) [inline lib/wallaby/phantom/driver.ex]: Keeping track of them in the session sounds like a great idea, much less clunky then a second function callback. they would all just go into a list the way they occured and could be accesses.

That people don't reassign the session shouldn't be a problem, it's immutable data structures and at some point people should know that and it's common to how session/conn work everywhere else as well.

File: lib/wallaby/phantom/driver.ex (hunk 11 above)
Comment (PragTob) [inline lib/wallaby/phantom/driver.ex]: huh that seems unexpected. Not a huge fans of code comments most of the time, but this would be a great place for a WHY comment (like the one you wrote up just now) so we don't stumble upon it along the road :)

File: lib/wallaby/browser.ex (hunk 8 above)
Comment (keathley) [inline lib/wallaby/browser.ex]: Is the string argument a requirement for the phantom driver? Or is that something that we should support at the Browser level? If its a part of the Browser api then we should have a guard in the Browser function. That way we can guard at the highest level possible.

File: lib/wallaby/phantom/driver.ex (hunk 9 above)
Comment (keathley) [inline lib/wallaby/phantom/driver.ex]: I agree. Thats something that we can test for and fix as a separate PR. If it does throw an exception then I _think_ it should crash the test with the correct stack trace.

File: lib/wallaby/phantom/driver.ex (hunk 10 above)
Comment (keathley) [inline lib/wallaby/phantom/driver.ex]: I actually like this approach much more as well. It allows us to chain the dialog functions without a clunky interface. We already do something similar with screenshots.

File: lib/wallaby/phantom/driver.ex (hunk 11 above)
Comment (keathley) [inline lib/wallaby/phantom/driver.ex]: Do these stack traces get picked up by the JS Logger? Or are they contained? Should we provide a more meaningful `{:ok, value} | {:error, error}` response back to the `Browser` module so that we can provide a better error message for the end user?

File: lib/wallaby/browser.ex (hunk 8 above)
Comment (padde) [inline lib/wallaby/browser.ex]: A `window.prompt` returns:

* `null` if dismissed
* `""` if accepted without input (and without a default value in place of course)

This means if we "accept" a prompt with a value of `null`, we actually dismiss it. There is no way to provide `null` or even non-string input to a prompt, so I'd say this should be handled on the Browser level.

File: lib/wallaby/phantom/driver.ex (hunk 10 above)
Comment (padde) [inline lib/wallaby/phantom/driver.ex]: @keathley what do you think about the caveat outlined in my code examples? Could be tricky to debug if you write it like in the second example.

We don't have any place to store session state except in the `session` struct, right? This change makes it the User's responsibility to reassign `session` to the updated struct that contains the messages.

If that's OK with you, I could start working on this. Should happen before the next release though, otherwise we'd immediately have incompatible changes in the release after that one. I'd be OK with directly putting it in this PR here.

Comment (PragTob) [inline lib/wallaby/phantom/driver.ex]: Users have to reassign the session all the time, so I think it's fine as it's fairly common practice. Also they don't HAVE TO reassign, they can just pipe it into a function that then looks it up and returns the session again. I use that a lot when writing my tests.

So it's fine. If you prefer, to do it here you can but the PR is already fairly large so reviews get cumbersome so my personal preference is for an extra PR personally :)

File: lib/wallaby/phantom/driver.ex (hunk 11 above)
Comment (padde) [inline lib/wallaby/phantom/driver.ex]: I'm not sure what you mean @keathley, but for clarification here's an example return value from `execute_phantom_script` when an error occurred:

```
IO.inspect execute_phantom_script(session, "boom")

{:ok,
 %{"line" => 1,
   "stack" => "anonymous\n_executePhantomJS@phantomjs://code/session.js:650:30\n_executePhantomJS@phantomjs://code/session_request_handler.js:863:68\n_handle@phantomjs://code/session_request_handler.js:166:30\n_reroute@phantomjs://code/request_handler.js:61:20\n_handle@phantomjs://code/router_request_handler.js:78:46"}}
```

The error message seems **not** to be included, just the line and stack trace. Any suggestions on how to improve this?

File: lib/wallaby/phantom/driver.ex (hunk 10 above)
Comment (padde) [inline lib/wallaby/phantom/driver.ex]: I was thinking this PR for a single reason: getting the API right before merging. How about I make a separate PR that uses this PR as base. You can review it in isolation and when it's good it can be merged into this PR. Once we're all set to release we can merge this PR with all the changes related to dialogs.

File: lib/wallaby/phantom/driver.ex (hunk 11 above)
Comment (keathley) [inline lib/wallaby/phantom/driver.ex]: So typically if there are javascript errors in the page those will get picked up by wallaby's internal JS logger. If you execute a script does it get picked up by that logger? I'm guessing that it doesn't.

Comment (padde) [inline lib/wallaby/phantom/driver.ex]: How can I best access the output JS Logger? If we use the "normal" `execute_script` function to trigger an error, it will not crash. But it returns something like this, which at least includes the original error message.


```
IO.inspect execute_script(session, "boom")

%{"class" => "org.openqa.selenium.WebDriverException",
  "message" => "{\"errorMessage\":\"Can't find variable: boom\",\"request\":{\"headers\":{\"Content-Length\":\"27\",\"Content-Type\":\"text/json, text/json\",\"Host\":\"localhost:51992\",\"User-Agent\":\"hackney/1.6.6\"},\"httpVersion\":\"1.1\",\"method\":\"POST\",\"post\":\"{\\\"script\\\":\\\"boom\\\",\\\"args\\\":[]}\",\"url\":\"/execute\",\"urlParsed\":{\"anchor\":\"\",\"query\":\"\",\"file\":\"execute\",\"directory\":\"/\",\"path\":\"/execute\",\"relative\":\"/execute\",\"port\":\"\",\"host\":\"\",\"password\":\"\",\"user\":\"\",\"userInfo\":\"\",\"authority\":\"\",\"protocol\":\"\",\"source\":\"/execute\",\"queryKey\":{},\"chunks\":[\"execute\"]},\"urlOriginal\":\"/session/5cb946a0-19c4-11e7-aef6-3973bb41a47b/execute\"}}",
  "screen" => "<Large Base64 encoded string ...>"}
```


Since both `execute_script` and `execute_phantom_script` internally use `check_logs!` I suppose that is what you mean with the internal JS Logger. AFAICS both these return values already come from the logger.

It would be really nice to raise an error in Wallaby if an exception is raised in  `execute_script` or `execute_phantom_script`.

File: lib/wallaby/phantom/driver.ex (hunk 10 above)
Comment (PragTob) [inline lib/wallaby/phantom/driver.ex]: it's fine if you prefer this PR, get the basic implementation in a separate commit then it's also rather easy to review. Thanks a lot :)

Comment (padde) [inline lib/wallaby/phantom/driver.ex]: @PragTob @keathley I am not so sure any more that this is worth it. The initial idea was that we wouldn't have to "interrupt" the pipeline in order to get the message later, but this approach kind of fails to live up to the expectations. I was trying out the approach just now and it turns out that instead of writing the test case like this:

```elixir
test "accept window.alert and get message", %{page: page} do
  message = accept_alert page, fn(p) ->
    click(p, Query.link("Alert"))
  end
  result =
    page
    |> find(Query.css("#result"))
    |> Element.text()
  assert message == "This is an alert!"
  assert result == "Alert accepted"
end
```

I now have to write it like this, because the modified `session` needs to be stored before we can continue:

```elixir
test "accept window.alert and get message", %{page: page} do
  page =
    page
    |> accept_alert(fn(p) ->
      click(p, Query.link("Alert"))
    end)
  result =
    page
    |> find(Query.css("#result"))
    |> Element.text()
  assert last_dialog_message(page) == "This is an alert!"
  assert result == "Alert accepted"
end
```

I am honestly wondering if this is any better than before, in particular if this is worth the additional complexity, having to keep track of messages in the session struct. Basically, this only simplifies the case you're using the finer grained `accept_alert` etc functions, but are not interested in the message.

It would be great to have some feedback from you before I actually dig in and implement this.

File: lib/wallaby/phantom/driver.ex (hunk 11 above)
Comment (keathley) [inline lib/wallaby/phantom/driver.ex]: Yes I agree. We can open a new issue for that.

File: lib/wallaby/phantom/driver.ex (hunk 10 above)
Comment (keathley) [inline lib/wallaby/phantom/driver.ex]: Don't worry about implementing this for now. Lets just return the alert text from `accept_alert` and the other helpers. Once the api has seen some serious use then we'll know better what direction we want to go.

Comment (padde) [inline lib/wallaby/phantom/driver.ex]: The pragmatic approach 😉 I'm fine leaving this as it is, thanks for the feedback guys.

**Comment by keathley** [comment]
@padde I'm going to do a final review of everything here, but is there anything else that you think we need to look at before we merge?

**Comment by padde** [comment]
@keathley I'd like to add the guard for the `input_value` in `accept_prompt` (see https://github.com/keathley/wallaby/pull/173#discussion_r109278665) before merging. Other than that we should be good to go.

**Comment by padde** [comment]
@keathley added the guard for the `input_value` with `accept_prompt/3`, should be good to go now I think.

**Comment by PragTob** [comment]
Okay I think this should be good to go now, besides - we might need it at work :D 

I'll try to dig through and file some issues we've mentioned.

**Comment by PragTob** [comment]
Ok just found one case where I really want to open an issue, if I missed any others please feel free tom open them :)
